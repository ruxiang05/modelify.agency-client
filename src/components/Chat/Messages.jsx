/* Uses methods and/or components from react, prop-types, react-router-dom and socket.io-client */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

class Messages extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      messages: [],
    };

    this.messagesEnd = React.createRef();
  }

  componentDidMount() {
    const { socket, chat } = this.props;
    const { messages } = chat;
    if (messages) {
      this.setState({ messages });
    }

    this.scrollToEnd();

    socket.on('receive_message', (data) => {
      if (chat._id === data.room) {
        this.addMessage(data);
      }
    });
  }

  componentDidUpdate() {
    this.scrollToEnd();
  }

  addMessage(data) {
    this.setState(prevState => ({ messages: [...prevState.messages, data] }));
  }

  scrollToEnd() {
    this.messagesEnd.current.scrollIntoView({ behaviour: 'smooth' });
  }

  render() {
    const { messages } = this.state;
    const { chat, user } = this.props;
    const { job } = chat;
    return (
      <>
        <div className="job-chat-card">
          <h2>{job.title}</h2>
          <Link
            to={{
              pathname: `/jobs/${job._id}`,
              state: job,
            }}
          >
View

          </Link>
        </div>
        <div className="messages">
          {messages.map(message => (
            <div
              key={messages.indexOf(message)}
              className={message.user === user._id ? 'message sent' : 'message received'}
            >
              <p>
                {message.message}
              </p>
            </div>
          ))}
          <div ref={this.messagesEnd} />
        </div>
      </>
    );
  }
}

Messages.propTypes = {
  socket: PropTypes.shape({}).isRequired,
  user: PropTypes.shape({}).isRequired,
  chat: PropTypes.shape({}).isRequired,
};


export default Messages;
