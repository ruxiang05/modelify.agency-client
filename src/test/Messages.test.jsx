/* Uses methods and/or components from react, socket.io-client and enzyme */
import React from 'react';
import socket from 'socket.io-client';
import { shallow } from 'enzyme';
import Messages from '../components/Chat/Messages';

let wrapper;

const props = {
  socket,
  chat: {
    _id: '2',
    messages: [{ message: 'hi', user: '3' }],
    job: {
      _id: '1',
      title: 'title',
    },
  },
  user: {
    _id: '3',
  },
};

Messages.prototype.scrollToEnd = jest.fn();
socket.on = jest.fn();

beforeEach(() => {
  wrapper = shallow(<Messages {...props} />);
});

describe('MessageForm', () => {
  it('should render MessageForm', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should add new messages', async () => {
    const data = {
      message: 'hi',
      user: '4',
    };
    wrapper.instance().addMessage(data);
    await wrapper.update();
    expect(wrapper.state().messages.length).toBe(2);
  });
});
