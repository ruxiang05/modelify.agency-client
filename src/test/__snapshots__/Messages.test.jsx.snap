// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm should render MessageForm 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Messages
    chat={
      Object {
        "_id": "2",
        "job": Object {
          "_id": "1",
          "title": "title",
        },
        "messages": Array [
          Object {
            "message": "hi",
            "user": "3",
          },
        ],
      }
    }
    socket={[MockFunction]}
    user={
      Object {
        "_id": "3",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="job-chat-card"
        >
          <h2>
            title
          </h2>
          <Link
            replace={false}
            to={
              Object {
                "pathname": "/jobs/1",
                "state": Object {
                  "_id": "1",
                  "title": "title",
                },
              }
            }
          >
            View
          </Link>
        </div>,
        <div
          className="messages"
        >
          <div
            className="message sent"
          >
            <p>
              hi
            </p>
          </div>
          <div />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              title
            </h2>,
            <Link
              replace={false}
              to={
                Object {
                  "pathname": "/jobs/1",
                  "state": Object {
                    "_id": "1",
                    "title": "title",
                  },
                }
              }
            >
              View
            </Link>,
          ],
          "className": "job-chat-card",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "title",
            },
            "ref": null,
            "rendered": "title",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "View",
              "replace": false,
              "to": Object {
                "pathname": "/jobs/1",
                "state": Object {
                  "_id": "1",
                  "title": "title",
                },
              },
            },
            "ref": null,
            "rendered": "View",
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <div
                className="message sent"
              >
                <p>
                  hi
                </p>
              </div>,
            ],
            <div />,
          ],
          "className": "messages",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": <p>
                hi
              </p>,
              "className": "message sent",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "hi",
              },
              "ref": null,
              "rendered": "hi",
              "type": "p",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="job-chat-card"
          >
            <h2>
              title
            </h2>
            <Link
              replace={false}
              to={
                Object {
                  "pathname": "/jobs/1",
                  "state": Object {
                    "_id": "1",
                    "title": "title",
                  },
                }
              }
            >
              View
            </Link>
          </div>,
          <div
            className="messages"
          >
            <div
              className="message sent"
            >
              <p>
                hi
              </p>
            </div>
            <div />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                title
              </h2>,
              <Link
                replace={false}
                to={
                  Object {
                    "pathname": "/jobs/1",
                    "state": Object {
                      "_id": "1",
                      "title": "title",
                    },
                  }
                }
              >
                View
              </Link>,
            ],
            "className": "job-chat-card",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "title",
              },
              "ref": null,
              "rendered": "title",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "View",
                "replace": false,
                "to": Object {
                  "pathname": "/jobs/1",
                  "state": Object {
                    "_id": "1",
                    "title": "title",
                  },
                },
              },
              "ref": null,
              "rendered": "View",
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <div
                  className="message sent"
                >
                  <p>
                    hi
                  </p>
                </div>,
              ],
              <div />,
            ],
            "className": "messages",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": <p>
                  hi
                </p>,
                "className": "message sent",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "hi",
                },
                "ref": null,
                "rendered": "hi",
                "type": "p",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
